src/transform/src/notices/mod.rs

```diff
@@ -0,0 +1,351 @@
+                $(
+                    #[doc = concat!("See [`", stringify!($ty), "`].")]
+                    $ty($ty),
+                )+
+            }
+
+            impl OptimizerNoticeApi for RawOptimizerNotice {
+                fn dependent_items(&self) -> BTreeSet<GlobalId> {
+                    match self {
+                        $(Self::$ty(notice) => notice.dependent_items(),)+
+                    }
+                }
+
+                fn fmt_message(&self, f: &mut Formatter<'_>, humanizer: &dyn ExprHumanizer) -> fmt::Result {
+                    match self {
+                        $(Self::$ty(notice) => notice.fmt_message(f, humanizer),)+
+                    }
+                }
+
+                fn fmt_hint(&self, f: &mut Formatter<'_>, humanizer: &dyn ExprHumanizer) -> fmt::Result {
+                    match self {
+                        $(Self::$ty(notice) => notice.fmt_hint(f, humanizer),)+
+                    }
+                }
+
+                fn fmt_action(&self, f: &mut Formatter<'_>, humanizer: &dyn ExprHumanizer) -> fmt::Result {
+                    match self {
+                        $(Self::$ty(notice) => notice.fmt_action(f, humanizer),)+
+                    }
+                }
+
+                fn action_kind(&self) -> ActionKind {
+                    match self {
+                        $(Self::$ty(notice) => notice.action_kind(),)+
+                    }
+                }
+            }
+
+            impl OptimizerNoticeKind {
+                /// Return a string representation for this optimizer notice
+                /// kind.
+                pub fn as_str(&self) -> &'static str {
+                    match self {
+                        $(Self::$ty => $name,)+
+                    }
+                }
+
+                /// A notice name, which will be applied as the label on the
+                /// metric that is counting notices labelled by notice kind.
+                pub fn metric_label(&self) -> &str {
+                    match self {
+                        $(Self::$ty => stringify!($ty),)+
+                    }
+                }
+            }
+
+            $(
+                impl From<$ty> for RawOptimizerNotice {
+                    fn from(value: $ty) -> Self {
+                        RawOptimizerNotice::$ty(value)
+                    }
+                }
+            )+
+        }
+    };
+}
+
+raw_optimizer_notices![
+    IndexTooWideForLiteralConstraints => "Index too wide for literal constraints",
+    IndexKeyEmpty => "Empty index key",
+];
+
+impl RawOptimizerNotice {
+    /// Turns the `OptimizerNotice` into a string, using the given `ExprHumanizer`.
+    pub fn to_string<'a>(&'a self, humanizer: &'a dyn ExprHumanizer) -> (String, String) {
+        (
+            self.message(humanizer).to_string(),
+            self.hint(humanizer).to_string(),
+        )
+    }
+
+    /// Turns a `Vec<OptimizerNotice>` into a String that can be used in EXPLAIN.
```

Comment:
Similarly, this returns a `Vec<String>`, not a `String`.

Reply:
Fixed the return type in the comment.